extras/markdowner.rb

```diff
@@ -7,49 +7,71 @@ def self.to_html(text, opts = {})
       return ""
     end
 
-    args = [ :smart, :autolink, :safelink, :filter_styles, :filter_html,
-      :strict ]
-    if !opts[:allow_images]
-      args.push :no_image
+    exts = [:tagfilter, :autolink, :strikethrough]
+    root = CommonMarker.render_doc(text.to_s, [:SMART], exts)
+
+    unless opts[:disable_profile_links]
+      walk_text_nodes(root) {|n| postprocess_text_node(n)}
     end
 
-    ng = Nokogiri::HTML(RDiscount.new(text.to_s, *args).to_html)
+    ng = Nokogiri::HTML(root.to_html([:SAFE], exts))
 
     # change <h1>, <h2>, etc. headings to just bold tags
     ng.css("h1, h2, h3, h4, h5, h6").each do |h|
       h.name = "strong"
     end
 
+    if !opts[:allow_images]
+      ng.css("img").remove
+    end
+
     # make links have rel=nofollow
     ng.css("a").each do |h|
-      h[:rel] = "nofollow"
+      h[:rel] = "nofollow" unless h[:href].starts_with?("/")
     end
 
-    # XXX: t.replace(tx) unescapes HTML, so disable for now.  this probably
-    # needs to split text into separate nodes and then replace the @username
-    # with a proper 'a' node
-if false
-    unless opts[:disable_profile_links]
-      # make @username link to that user's profile
-      ng.search("//text()").each do |t|
-        if t.parent && t.parent.name.downcase == "a"
-          # don't replace inside <a>s
-          next
-        end
-
-        tx = t.text.gsub(/\B\@([\w\-]+)/) do |u|
-          if User.exists?(:username => u[1 .. -1])
-            "<a href=\"/u/#{u[1 .. -1]}\">#{u}</a>"
-          else
-            u
-          end
-        end
-
-        t.replace(tx)
-      end
+    ng.at_css("body").inner_html
+  end
+
+  def self.walk_text_nodes(node, &block)
+    return if node.type == :link
+    return block.call(node) if node.type == :text
+
+    node.each do |child|
+      walk_text_nodes(child, &block)
     end
-end
+  end
 
-    ng.at_css("body").inner_html
+  def self.postprocess_text_node(node)
+    while node
+      return unless node.string_content =~ /\B(@[\w\-]+)/
```

Comment:
It might be worth moving the username validation regex in `User` out to a constant and checking against that.